addons/advanced_ballistics/functions/fnc_calculateBarrelLengthVelocityShift.sqf

```diff
@@ -10,46 +10,62 @@
-_barrelLength        = _this select 0;
-_muzzleVelocityTable = _this select 1;
-_barrelLengthTable   = _this select 2;
-_muzzleVelocity      = _this select 3;
+scopeName "main";
 
+private ["_muzzleVelocityTableCount", "_barrelLengthTableCount", "_lowerDataIndex",
+    "_upperDataIndex", "_lowerBarrelLength", "_upperBarrelLength", "_lowerMuzzleVelocity",
+    "_upperMuzzleVelocity", "_interpolationRatio"];
+params ["_barrelLength", "_muzzleVelocityTable", "_barrelLengthTable", "_muzzleVelocity"];
+
+// If barrel length is not defined, then there is no point in calculating muzzle velocity
 if (_barrelLength == 0) exitWith { 0 };
-if (count _muzzleVelocityTable != count _barrelLengthTable) exitWith { 0 };
-if (count _muzzleVelocityTable == 0 || count _barrelLengthTable == 0) exitWith { 0 };
-if (count _muzzleVelocityTable == 1) exitWith { (_muzzleVelocityTable select 0) - _muzzleVelocity };
 
-_lowerIndex = 0;
-_upperIndex = (count _barrelLengthTable) - 1;
+_muzzleVelocityTableCount = count _muzzleVelocityTable;
+_barrelLengthTableCount = count _barrelLengthTable;
 
-if (_barrelLength <= (_barrelLengthTable select _lowerIndex)) exitWith { (_muzzleVelocityTable select _lowerIndex) - _muzzleVelocity };
-if (_barrelLength >= (_barrelLengthTable select _upperIndex)) exitWith { (_muzzleVelocityTable select _upperIndex) - _muzzleVelocity };
+// Exit if tables are different sizes, have no elements or have only one element
+if (_muzzleVelocityTableCount != _barrelLengthTableCount) exitWith { 0 };
+if (_muzzleVelocityTableCount == 0 || _barrelLengthTableCount == 0) exitWith { 0 };
+if (_muzzleVelocityTableCount == 1) exitWith { (_muzzleVelocityTable select 0) - _muzzleVelocity };
 
-for "_i" from 0 to (count _barrelLengthTable) - 1 do {
-    if (_barrelLength >= _barrelLengthTable select _i) then {
-        _lowerIndex = _i;
-    };
+// If we have the precise barrel length value, return result immediately
+if (_barrelLength in _barrelLengthTable) exitWith {
+    _muzzleVelocityTable select (_barrelLengthTable find _barrelLength) - _muzzleVelocity
 };
-for "_i" from (count _barrelLengthTable) - 1 to 0 step -1 do {
-    if (_barrelLength <= _barrelLengthTable select _i) then {
-        _upperIndex = _i;
+
+// Limit values to lower and upper bound of data we have available
+if (_barrelLength <= (_barrelLengthTable select 0)) exitWith { (_muzzleVelocityTable select 0) - _muzzleVelocity };
+if (_barrelLength >= (_barrelLengthTable select _barrelLengthTableCount - 1)) exitWith { (_muzzleVelocityTable select _barrelLengthTableCount - 1) - _muzzleVelocity };
+
+// Find closest bordering values for barrel length
+{
+    if (_barrelLength >= _x) then {
+        _lowerDataIndex = _forEachIndex;
+        _upperDataIndex = _lowerDataIndex + 1;
+        breakTo "main";
     };
-};
+} forEach _barrelLengthTable;
```

Comment:
Yeah, but like you said: https://github.com/acemod/ACE3/pull/2015#discussion_r37706755
The corner cases are already handled when the execution path reaches this section.